File: app/src/main/java/tpcreative/co/qrscanner/common/view/crop/CropImageActivity.java
package tpcreative.co.qrscanner.common.view.crop;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.BitmapRegionDecoder;
import android.graphics.Matrix;
import android.graphics.Rect;
import android.graphics.RectF;
import android.net.Uri;
import android.opengl.GLES10;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.provider.MediaStore;
import android.view.View;
import android.view.Window;
import android.widget.FrameLayout;
import androidx.core.content.ContextCompat;
import com.google.gson.Gson;
import com.google.zxing.BinaryBitmap;
import com.google.zxing.ChecksumException;
import com.google.zxing.FormatException;
import com.google.zxing.LuminanceSource;
import com.google.zxing.MultiFormatReader;
import com.google.zxing.NotFoundException;
import com.google.zxing.RGBLuminanceSource;
import com.google.zxing.Reader;
import com.google.zxing.Result;
import com.google.zxing.common.HybridBinarizer;
import java.io.IOException;
import java.io.InputStream;
import java.util.concurrent.Callable;
import java.util.concurrent.CountDownLatch;
import butterknife.BindView;
import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.schedulers.Schedulers;
import tpcreative.co.qrscanner.R;
import tpcreative.co.qrscanner.common.Utils;

public class CropImageActivity extends MonitoredActivity implements CropImageView.ListenerState {
    private static final String TAG = CropImageActivity.class.getSimpleName();
    private static final int SIZE_DEFAULT = 2048;
    private static final int SIZE_LIMIT = 4096;
    private final Handler handler  = new Handler(Looper.getMainLooper());
    CompositeDisposable compositeDisposable  = new CompositeDisposable();;
    private int aspectX;
    private int aspectY;
    // Output image
    private int maxX;
    private int maxY;
    private int exifRotation;
    private Uri sourceUri;
    private boolean isSaving;
    private int sampleSize;
    private RotateBitmap rotateBitmap;
    private CropImageView imageView;
    private HighlightView cropView;
    private boolean isProgressing ;
    @BindView(R.id.btn_done)
    FrameLayout layoutDone;

    @Override
    public void onCreate(Bundle icicle) {
        requestWindowFeature(Window.FEATURE_NO_TITLE);
        super.onCreate(icicle);
        setupViews();
        loadInput();
        if (rotateBitmap == null) {
            finish();
            return;
        }
        startCrop();
        layoutDone.setEnabled(false);
        layoutDone.setBackgroundColor(ContextCompat.getColor(this,R.color.colorAccent));
    }

    private void setupViews() {
        setContentView(R.layout.crop_activity_crop);
        imageView = (CropImageView) findViewById(R.id.crop_image);
        imageView.context = this;
        imageView.setRecycler(new ImageViewTouchBase.Recycler() {
            @Override
            public void recycle(Bitmap b) {
                b.recycle();
                System.gc();
            }
        });
        findViewById(R.id.btn_cancel).setOnClickListener(new View.OnClickListener() {
            public void onClick(View v) {
                setResult(RESULT_CANCELED);
                finish();
            }
        });

        findViewById(R.id.btn_done).setOnClickListener(new View.OnClickListener() {
            public void onClick(View v) {
               finish();
            }
        });
    }

    private void loadInput() {
        Intent intent = getIntent();
        Bundle extras = intent.getExtras();
        if (extras != null) {
            aspectX = extras.getInt(Crop.Extra.ASPECT_X);
            aspectY = extras.getInt(Crop.Extra.ASPECT_Y);
            maxX = extras.getInt(Crop.Extra.MAX_X);
            maxY = extras.getInt(Crop.Extra.MAX_Y);
        }
        sourceUri = intent.getData();
        if (sourceUri != null) {
            exifRotation = CropUtil.getExifRotation(CropUtil.getFromMediaUri(this, getContentResolver(), sourceUri));
            InputStream is = null;
            try {
                sampleSize = calculateBitmapSampleSize(sourceUri);
                is = getContentResolver().openInputStream(sourceUri);
                BitmapFactory.Options option = new BitmapFactory.Options();
                option.inSampleSize = sampleSize;
                rotateBitmap = new RotateBitmap(BitmapFactory.decodeStream(is, null, option), exifRotation);
            } catch (IOException e) {
                Log.e("Error reading image: " + e.getMessage(), e);
                setResultException(e);
            } catch (OutOfMemoryError e) {
                Log.e("OOM reading image: " + e.getMessage(), e);
                setResultException(e);
            } finally {
                CropUtil.closeSilently(is);
            }
        }
    }

    private int calculateBitmapSampleSize(Uri bitmapUri) throws IOException {
        InputStream is = null;
        BitmapFactory.Options options = new BitmapFactory.Options();
        options.inJustDecodeBounds = true;
        try {
            is = getContentResolver().openInputStream(bitmapUri);
            BitmapFactory.decodeStream(is, null, options); // Just get image size
        } finally {
            CropUtil.closeSilently(is);
        }

        int maxSize = getMaxImageSize();
        int sampleSize = 1;
        while (options.outHeight / sampleSize > maxSize || options.outWidth / sampleSize > maxSize) {
            sampleSize = sampleSize << 1;
        }
        return sampleSize;
    }

    private int getMaxImageSize() {
        int textureLimit = getMaxTextureSize();
        if (textureLimit == 0) {
            return SIZE_DEFAULT;
        } else {
            return Math.min(textureLimit, SIZE_LIMIT);
        }
    }

    private int getMaxTextureSize() {
        // The OpenGL texture size is the maximum size that can be drawn in an ImageView
        int[] maxSize = new int[1];
        GLES10.glGetIntegerv(GLES10.GL_MAX_TEXTURE_SIZE, maxSize, 0);
        return maxSize[0];
    }

    private void startCrop() {
        if (isFinishing()) {
            return;
        }
        imageView.setListenerState(this);
        imageView.setImageRotateBitmapResetBase(rotateBitmap, true);
        CropUtil.startBackgroundJob(this, null, getResources().getString(R.string.crop__wait),
                new Runnable() {
                    public void run() {
                        final CountDownLatch latch = new CountDownLatch(1);
                        handler.post(new Runnable() {
                            public void run() {
                                if (imageView.getScale() == 1F) {
                                    imageView.center();
                                }
                                latch.countDown();
                            }
                        });
                        try {
                            latch.await();
                        } catch (InterruptedException e) {
                            throw new RuntimeException(e);
                        }
                        new Cropper().crop();
                    }
                }, handler
        );
    }

    private class Cropper {
        private void makeDefault() {
            if (rotateBitmap == null) {
                return;
            }
            HighlightView hv = new HighlightView(imageView);
            final int width = rotateBitmap.getWidth();
            final int height = rotateBitmap.getHeight();
            Rect imageRect = new Rect(0, 0, width, height);
            // Make the default size about 4/5 of the width or height
            int cropWidth = Math.min(width, height) * 4 / 5;
            @SuppressWarnings("SuspiciousNameCombination")
            int cropHeight = cropWidth;
            if (aspectX != 0 && aspectY != 0) {
                if (aspectX > aspectY) {
                    cropHeight = cropWidth * aspectY / aspectX;
                } else {
                    cropWidth = cropHeight * aspectX / aspectY;
                }
            }
            int x = (width - cropWidth) / 2;
            int y = (height - cropHeight) / 2;
            RectF cropRect = new RectF(x, y, x + cropWidth, y + cropHeight);
            hv.setup(imageView.getUnrotatedMatrix(), imageRect, cropRect, aspectX != 0 && aspectY != 0);
            imageView.add(hv);
        }

        public void crop() {
            handler.post(new Runnable() {
                public void run() {
                    makeDefault();
                    imageView.invalidate();
                    if (imageView.highlightViews.size() == 1) {
                        cropView = imageView.highlightViews.get(0);
                        cropView.setFocus(true);
                    }
                }
            });
        }
    }


    private Bitmap decodeRegionCrop(Rect rect, int outWidth, int outHeight) {
        // Release memory now
//        clearImageView();
        InputStream is = null;
        Bitmap croppedImage = null;
        try {
            is = getContentResolver().openInputStream(sourceUri);
            BitmapRegionDecoder decoder = BitmapRegionDecoder.newInstance(is, false);
            final int width = decoder.getWidth();
            final int height = decoder.getHeight();
            if (exifRotation != 0) {
                // Adjust crop area to account for image rotation
                Matrix matrix = new Matrix();
                matrix.setRotate(-exifRotation);

                RectF adjusted = new RectF();
                matrix.mapRect(adjusted, new RectF(rect));

                // Adjust to account for origin at 0,0
                adjusted.offset(adjusted.left < 0 ? width : 0, adjusted.top < 0 ? height : 0);
                rect = new Rect((int) adjusted.left, (int) adjusted.top, (int) adjusted.right, (int) adjusted.bottom);
            }
            try {
                croppedImage = decoder.decodeRegion(rect, new BitmapFactory.Options());
                if (croppedImage != null && (rect.width() > outWidth || rect.height() > outHeight)) {
                    Matrix matrix = new Matrix();
                    matrix.postScale((float) outWidth / rect.width(), (float) outHeight / rect.height());
                    croppedImage = Bitmap.createBitmap(croppedImage, 0, 0, croppedImage.getWidth(), croppedImage.getHeight(), matrix, true);
                }
            } catch (IllegalArgumentException e) {
                // Rethrow with some extra information
                throw new IllegalArgumentException("Rectangle " + rect + " is outside of the image ("
                        + width + "," + height + "," + exifRotation + ")", e);
            }
        } catch (IOException e) {
            Log.e("Error cropping image: " + e.getMessage(), e);
            setResultException(e);
        } catch (OutOfMemoryError e) {
            Log.e("OOM cropping image: " + e.getMessage(), e);
            setResultException(e);
        } finally {
            CropUtil.closeSilently(is);
        }
        return croppedImage;
    }

    private void clearImageView() {
        imageView.clear();
        if (rotateBitmap != null) {
            rotateBitmap.recycle();
        }
        System.gc();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (rotateBitmap != null) {
            rotateBitmap.recycle();
        }
        compositeDisposable.dispose();
    }

    @Override
    public boolean onSearchRequested() {
        return false;
    }

    public boolean isSaving() {
        return isSaving;
    }

    private void setResultUri(Uri uri) {
        setResult(RESULT_OK, new Intent().putExtra(MediaStore.EXTRA_OUTPUT, uri));
    }

    private void setResultEncode(Result encode) {
        setResult(RESULT_OK, new Intent().putExtra(Crop.REQUEST_DATA,new Gson().toJson(encode)));
    }

    private void setResultException(Throwable throwable) {
        setResult(Crop.RESULT_ERROR, new Intent().putExtra(Crop.Extra.ERROR, throwable));
    }

    @Override
    public boolean isProgressingCropImage() {
        return isProgressing;
    }

    @Override
    public void onRequestCropImage() {
        Utils.Log(TAG,"onRequestCropImage");
        handleCropping();
    }

    public void handleCropping(){
        if (cropView == null || isSaving) {
            return;
        }
        isSaving = true;
        isProgressing = true;
        Bitmap croppedImage;
        Rect r = cropView.getScaledCropRect(sampleSize);
        int width = r.width();
        int height = r.height();
        int outWidth = width;
        int outHeight = height;
        if (maxX > 0 && maxY > 0 && (width > maxX || height > maxY)) {
            float ratio = (float) width / (float) height;
            if ((float) maxX / (float) maxY > ratio) {
                outHeight = maxY;
                outWidth = (int) ((float) maxY * ratio + .5f);
            } else {
                outWidth = maxX;
                outHeight = (int) ((float) maxX / ratio + .5f);
            }
        }
        try {
            croppedImage = decodeRegionCrop(r, outWidth, outHeight);
        } catch (IllegalArgumentException e) {
            setResultException(e);
            finish();
            return;
        }
        onRenderCode(croppedImage);
    }

    public void onRenderCode(final Bitmap bitmap){
        compositeDisposable.add(Observable.fromCallable(new Callable<String>() {
            @Override
            public String call() throws Exception {
                try{
                    if(bitmap==null){
                        isSaving = false;
                        isProgressing = false;
                        return "";
                    }
                    int[] intArray = new int[bitmap.getWidth()*bitmap.getHeight()];
                    bitmap.getPixels(intArray, 0, bitmap.getWidth(), 0, 0, bitmap.getWidth(), bitmap.getHeight());
                    LuminanceSource source = new RGBLuminanceSource(bitmap.getWidth(), bitmap.getHeight(), intArray);
                    BinaryBitmap mBitmap = new BinaryBitmap(new HybridBinarizer(source));
                    Reader reader = new MultiFormatReader();
                    try {
                        Result result = reader.decode(mBitmap);
                        Utils.Log(TAG,"This is type of qrcode");
                        return new Gson().toJson(result);
                    }
                    catch (NotFoundException  | ChecksumException e){
                        e.printStackTrace();
                        Utils.Log(TAG,"Do not recognize qrcode type");
                        return "";
                    }
                }
                catch (FormatException e){
                    e.printStackTrace();
                    Utils.Log(TAG,"Do not recognize qrcode type");
                    return "";
                }
            }
        }).subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread()).subscribe(response ->{
                    final Result mResult = new Gson().fromJson(response,Result.class);
                    if (mResult!=null ){
                        isSaving = false;
                        isProgressing = false;
                        layoutDone.setEnabled(true);
                        layoutDone.setBackgroundColor(ContextCompat.getColor(this,R.color.colorPrimary));
                        setResultEncode(mResult);
                    }else{
                        isSaving = false;
                        isProgressing = false;
                        layoutDone.setEnabled(false);
                        layoutDone.setBackgroundColor(ContextCompat.getColor(this,R.color.colorAccent));
                    }
                }));

        Utils.Log(TAG,"onRenderCode");
    }

    @Override
    public void onBackPressed() {
        setResult(RESULT_CANCELED);
        finish();
        super.onBackPressed();
    }


}

File: app/src/main/java/tpcreative/co/qrscanner/common/view/crop/CropImageActivity.kt
package tpcreative.co.qrscanner.common.view.crop
import android.app.Activity
import android.content.Intent
import android.graphics.*
import android.net.Uri
import android.opengl.GLES10
import android.os.*
import android.provider.MediaStore
import android.view.*
import android.window.OnBackInvokedDispatcher
import androidx.activity.OnBackPressedCallback
import androidx.activity.result.ActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.core.content.ContextCompat
import com.google.gson.Gson
import com.google.zxing.*
import com.google.zxing.client.result.*
import com.google.zxing.common.HybridBinarizer
import kotlinx.android.synthetic.main.crop_activity_crop.*
import kotlinx.android.synthetic.main.crop_layout_done_cancel.*
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import tpcreative.co.qrscanner.R
import tpcreative.co.qrscanner.common.*
import tpcreative.co.qrscanner.common.extension.onGeneralParse
import tpcreative.co.qrscanner.common.extension.parcelable
import tpcreative.co.qrscanner.common.view.crop.Crop.Extra
import tpcreative.co.qrscanner.common.view.crop.CropImageView.ListenerState
import tpcreative.co.qrscanner.model.GeneralModel
import tpcreative.co.qrscanner.model.EnumFragmentType
import tpcreative.co.qrscanner.model.EnumImplement
import tpcreative.co.qrscanner.ui.scannerresult.ScannerResultActivity
import java.io.IOException
import java.io.InputStream
import java.util.*
import java.util.concurrent.CountDownLatch

internal class CropImageActivity : MonitoredActivity(), ListenerState {
    private val handler: Handler = Handler(Looper.getMainLooper())
    private var aspectX = 0
    private var aspectY = 0

    // Output image
    private var maxX = 0
    private var maxY = 0
    private var exifRotation = 0
    private var sourceUri: Uri? = null
    private var isSaving = false
    private var sampleSize = 0
    private var rotateBitmap: RotateBitmap? = null
    private var imageView: CropImageView? = null
    private var cropView: HighlightView? = null
    private var isProgressing = false
    private var isShareIntent = false
    private var mCreate : GeneralModel? = null
    public override fun onCreate(icicle: Bundle?) {
        requestWindowFeature(Window.FEATURE_NO_TITLE)
        super.onCreate(icicle)
        setupViews()
        onHandlerIntent()
        if (rotateBitmap == null) {
            finish()
            return
        }
        startCrop()
        btn_done.isEnabled = false
        btn_done.setBackgroundColor(ContextCompat.getColor(this, R.color.colorAccent))
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            onBackInvokedDispatcher.registerOnBackInvokedCallback(
                OnBackInvokedDispatcher.PRIORITY_DEFAULT
            ) {
                setResult(RESULT_CANCELED)
                finish()
            }
        } else {
            onBackPressedDispatcher.addCallback(
                this,
                object : OnBackPressedCallback(true) {
                    override fun handleOnBackPressed() {
                        setResult(RESULT_CANCELED)
                        finish()
                    }
                })
        }
        //onHandlerIntent()
    }

    private fun setupViews() {
        setContentView(R.layout.crop_activity_crop)
        imageView = findViewById<View?>(R.id.crop_image) as CropImageView
        imageView?.mContext = this
        imageView?.setRecycler(object : ImageViewTouchBase.Recycler {
            override fun recycle(b: Bitmap?) {
                b?.recycle()
                System.gc()
            }
        })
        btn_cancel.setOnClickListener {
            setResult(RESULT_CANCELED)
            finish()
        }
       btn_done.setOnClickListener {
           if (isShareIntent){
               onDoNavigation()
           }
           finish()
       }
    }

    /*Share File To QRScanner*/
    private fun onHandlerIntent() {
        try {
            val action = intent?.action
            val type = intent?.type
            val extras = intent.extras
            if (extras != null) {
                aspectX = extras.getInt(Extra.ASPECT_X)
                aspectY = extras.getInt(Extra.ASPECT_Y)
                maxX = extras.getInt(Extra.MAX_X)
                maxY = extras.getInt(Extra.MAX_Y)
                Utils.Log(TAG,"$aspectX $aspectY $maxX $maxY")
            }
            if (Intent.ACTION_SEND == action && type != null) {
                sourceUri = handleSendSingleItem()
                isShareIntent = true
            }else{
                sourceUri = intent.data
                isShareIntent = false
            }
        } catch (e: Exception) {
            Utils.onDropDownAlert(this, getString(R.string.error_occurred_importing))
            e.printStackTrace()
        }
        loadInput()
    }

    private fun handleSendSingleItem()  : Uri?{
        try {
            val imageUri = intent?.parcelable<Parcelable>(Intent.EXTRA_STREAM) as Uri?
            if (imageUri != null) {
                return imageUri
            } else {
                Utils.onDropDownAlert(this, getString(R.string.can_not_support_this_format))
            }
        } catch (e: Exception) {
            Utils.onDropDownAlert(this, getString(R.string.error_occurred_importing))
            e.printStackTrace()
        }
        return null
    }

    private fun onParseData(result: Result?){
        try {
            result?.let {mResult->
                val parsedResult = ResultParser.parseResult(result)
                val create = Utils.onGeneralParse(mResult,GeneralModel::class)
                create.enumImplement = EnumImplement.VIEW
                create.fragmentType = EnumFragmentType.SCANNER
                create.barcodeFormat = BarcodeFormat.QR_CODE.name
                if (mResult.barcodeFormat != null) {
                    create.barcodeFormat = mResult.barcodeFormat.name
                }
                mCreate = create
                tvFormatType.text = parsedResult.type.name
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private fun onDoNavigation(){
        scanForResult.launch(Navigator.onResultView(this, mCreate, ScannerResultActivity::class.java))
    }

    private val scanForResult = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result: ActivityResult ->
        if (result.resultCode == Activity.RESULT_OK) {
            Utils.Log(TAG,"Okay")
        }
    }

    private fun loadInput() {
        if (sourceUri != null) {
            exifRotation = CropUtil.getExifRotation(CropUtil.getFromMediaUri(this, contentResolver, sourceUri))
            var mInputStream: InputStream? = null
            try {
                sampleSize = calculateBitmapSampleSize(sourceUri)
                mInputStream = contentResolver.openInputStream(sourceUri!!)
                val option = BitmapFactory.Options()
                option.inSampleSize = sampleSize
                rotateBitmap = RotateBitmap(BitmapFactory.decodeStream(mInputStream, null, option), exifRotation)
            } catch (e: IOException) {
                Log.e("Error reading image: " + e.message, e)
                setResultException(e)
            } catch (e: OutOfMemoryError) {
                Log.e("OOM reading image: " + e.message, e)
                setResultException(e)
            } finally {
                CropUtil.closeSilently(mInputStream)
            }
        }
    }

    @Throws(IOException::class)
    private fun calculateBitmapSampleSize(bitmapUri: Uri?): Int {
        var mInputStream: InputStream? = null
        val options = BitmapFactory.Options()
        options.inJustDecodeBounds = true
        try {
            mInputStream = bitmapUri?.let { contentResolver.openInputStream(it) }
            BitmapFactory.decodeStream(mInputStream, null, options) // Just get image size
        } finally {
            CropUtil.closeSilently(mInputStream)
        }
        val maxSize = getMaxImageSize()
        var sampleSize = 1
        while (options.outHeight / sampleSize > maxSize || options.outWidth / sampleSize > maxSize) {
            sampleSize = sampleSize shl 1
        }
        return sampleSize
    }

    private fun getMaxImageSize(): Int {
        val textureLimit = getMaxTextureSize()
        return if (textureLimit == 0) {
            SIZE_DEFAULT
        } else {
            textureLimit.coerceAtMost(SIZE_LIMIT)
        }
    }

    private fun getMaxTextureSize(): Int {
        // The OpenGL texture size is the maximum size that can be drawn in an ImageView
        val maxSize = IntArray(1)
        GLES10.glGetIntegerv(GLES10.GL_MAX_TEXTURE_SIZE, maxSize, 0)
        return maxSize[0]
    }

    private fun startCrop() {
        if (isFinishing) {
            return
        }
        imageView?.setListenerState(this)
        imageView?.setImageRotateBitmapResetBase(rotateBitmap, true)
        CropUtil.startBackgroundJob(this, null, resources.getString(R.string.crop__wait),
                {
                    val latch = CountDownLatch(1)
                    handler.post(Runnable {
                        if (imageView?.getScale() == 1f) {
                            imageView?.center()
                        }
                        latch.countDown()
                    })
                    try {
                        latch.await()
                    } catch (e: InterruptedException) {
                        throw RuntimeException(e)
                    }
                    Cropper().crop()
                }, handler
        )
    }

    private inner class Cropper {
        private fun makeDefault() {
            if (rotateBitmap == null) {
                return
            }
            val hv = HighlightView(imageView)
            val width = rotateBitmap?.getWidth() ?:0
            val height = rotateBitmap?.getHeight() ?:0
            val imageRect = Rect(0, 0, width, height)
            // Make the default size about 4/5 of the width or height
            var cropWidth = width.coerceAtMost(height) * 4 / 5
            var cropHeight = cropWidth
            if (aspectX != 0 && aspectY != 0) {
                if (aspectX > aspectY) {
                    cropHeight = cropWidth * aspectY / aspectX
                } else {
                    cropWidth = cropHeight * aspectX / aspectY
                }
            }
            val x = (width - cropWidth) / 2
            val y = (height - cropHeight) / 2
            var cropRect = RectF(x.toFloat(), y.toFloat(), x + cropWidth.toFloat(), y + cropHeight.toFloat())
            if (width>height){
                 //Detect is barcode
                 cropRect = RectF(1f, 1f, width.toFloat() - 1, height.toFloat() - 1)
            }
            hv.setup(imageView?.getUnrotatedMatrix(), imageRect, cropRect, aspectX != 0 && aspectY != 0)
            imageView?.add(hv)
            Utils.Log(TAG,"with: $width, height: $height")
        }

        fun crop() {
            handler.post(Runnable {
                makeDefault()
                imageView?.invalidate()
                if (imageView?.highlightViews?.size == 1) {
                    cropView = imageView?.highlightViews?.get(0)
                    cropView?.setFocus(true)
                }
            })
        }
    }

    private fun decodeRegionCrop(rect: Rect?, outWidth: Int, outHeight: Int): Bitmap? {
        // Release memory now
//        clearImageView();
        var rect = rect
        var mInput: InputStream? = null
        var croppedImage: Bitmap? = null
        try {
            mInput = sourceUri?.let { contentResolver.openInputStream(it) }
            val decoder = mInput?.let { bitmapRecognize(it) }
            val width = decoder?.width?.toFloat()
            val height = decoder?.height?.toFloat()
            if (exifRotation != 0) {
                // Adjust crop area to account for image rotation
                val matrix = Matrix()
                matrix.setRotate(-exifRotation.toFloat())
                val adjusted = RectF()
                matrix.mapRect(adjusted, RectF(rect))
                // Adjust to account for origin at 0,0
                (if (adjusted.left < 0) width else 0.toFloat())?.let { (if (adjusted.top < 0) height else 0.toFloat())?.let { it1 ->
                    adjusted.offset(it,
                        it1
                    )
                } }
                rect = Rect(adjusted.left.toInt(), adjusted.top.toInt(), adjusted.right.toInt(), adjusted.bottom.toInt())
            }
            try {
                croppedImage = decoder?.decodeRegion(rect, BitmapFactory.Options())
                if (croppedImage != null && (rect?.width()!! > outWidth || rect.height() > outHeight)) {
                    val matrix = Matrix()
                    matrix.postScale(outWidth.toFloat() / rect.width(), outHeight.toFloat() / rect.height())
                    croppedImage = Bitmap.createBitmap(croppedImage, 0, 0, croppedImage.width, croppedImage.height, matrix, true)
                }
            } catch (e: IllegalArgumentException) {
                // Rethrow with some extra information
                throw IllegalArgumentException("Rectangle " + rect + " is outside of the image ("
                        + width + "," + height + "," + exifRotation + ")", e)
            }
        } catch (e: IOException) {
            Log.e("Error cropping image: " + e.message, e)
            setResultException(e)
        } catch (e: OutOfMemoryError) {
            Log.e("OOM cropping image: " + e.message, e)
            setResultException(e)
        } finally {
            CropUtil.closeSilently(mInput)
        }
        return croppedImage
    }

    private fun clearImageView() {
        imageView?.clear()
        if (rotateBitmap != null) {
            rotateBitmap?.recycle()
        }
        System.gc()
    }

    override fun onDestroy() {
        super.onDestroy()
        if (rotateBitmap != null) {
            rotateBitmap?.recycle()
        }
    }

    override fun onSearchRequested(): Boolean {
        return false
    }

    fun isSaving(): Boolean {
        return isSaving
    }

    private fun setResultUri(uri: Uri?) {
        setResult(RESULT_OK, Intent().putExtra(MediaStore.EXTRA_OUTPUT, uri))
    }

    private fun setResultEncode(encode: Result?) {
        setResult(RESULT_OK, Intent().putExtra(Crop.REQUEST_DATA, Gson().toJson(encode)))
    }

    private fun setResultException(throwable: Throwable?) {
        setResult(Crop.RESULT_ERROR, Intent().putExtra(Extra.ERROR, throwable))
    }

    override fun isProgressingCropImage(): Boolean {
        return isProgressing
    }

    override fun onRequestCropImage() {
        Utils.Log(TAG, "onRequestCropImage")
        handleCropping()
    }

    private fun handleCropping() {
        if (cropView == null || isSaving) {
            return
        }
        isSaving = true
        isProgressing = true
        val croppedImage: Bitmap?
        val r = cropView?.getScaledCropRect(sampleSize.toFloat())
        val width = r?.width()?.toFloat() ?:0F
        val height = r?.height()?.toFloat() ?:0F
        var outWidth = width
        var outHeight = height
        if (maxX > 0 && maxY > 0 && (width > maxX || height > maxY)) {
            val ratio = width / height
            if (maxX.toFloat() / maxY.toFloat() > ratio) {
                outHeight = maxY.toFloat()
                outWidth = (maxY.toFloat() * ratio + .5f)
            } else {
                outWidth = maxX.toFloat()
                outHeight = (maxX.toFloat() / ratio + .5f)
            }
        }
        croppedImage = try {
            decodeRegionCrop(r, outWidth.toInt(), outHeight.toInt())
        } catch (e: IllegalArgumentException) {
            setResultException(e)
            finish()
            return
        }
        onRenderCode(croppedImage)
    }

    private fun onRenderCode(bitmap: Bitmap?) {
        CoroutineScope(Dispatchers.Main).launch {
            try {
                if (bitmap == null) {
                    isSaving = false
                    isProgressing = false
                    return@launch
                }
                val intArray = IntArray(bitmap.width * bitmap.height)
                bitmap.getPixels(intArray, 0, bitmap.width, 0, 0, bitmap.width, bitmap.height)
                val source: LuminanceSource = RGBLuminanceSource(bitmap.width, bitmap.height, intArray)
                val mBitmap = BinaryBitmap(HybridBinarizer(source))
                Utils.Log(TAG,"width ${bitmap.width} height ${bitmap.height}")
                val reader: Reader = MultiFormatReader()
                try {
                  var mResult : Result? = null
                    try {
                        mResult = reader.decode(mBitmap)
                    }catch (e : Exception){
                        e.printStackTrace()
                        try {
                            mResult = reader.decode(mBitmap,addHint())
                        }catch (e: Exception){
                            e.printStackTrace()
                        }
                    }
                    if (mResult != null) {
                        isSaving = false
                        isProgressing = false
                        btn_done.isEnabled = true
                        btn_done.setBackgroundColor(ContextCompat.getColor(this@CropImageActivity, R.color.colorPrimary))
                        setResultEncode(mResult)
                        onParseData(mResult)
                    } else {
                        isSaving = false
                        isProgressing = false
                        btn_done.isEnabled = false
                        btn_done.setBackgroundColor(ContextCompat.getColor(this@CropImageActivity, R.color.colorAccent))
                        tvFormatType.text = ""
                    }
                } catch (e: NotFoundException) {
                    e.printStackTrace()
                    Utils.Log(TAG, "Do not recognize qrcode type ${e.message}")
                    isSaving = false
                    isProgressing = false
                    btn_done.isEnabled = false
                    btn_done.setBackgroundColor(ContextCompat.getColor(this@CropImageActivity, R.color.colorAccent))
                } catch (e: ChecksumException) {
                    e.printStackTrace()
                    Utils.Log(TAG, "Do not recognize qrcode type ChecksumException")
                    isSaving = false
                    isProgressing = false
                    btn_done.isEnabled = false
                    btn_done.setBackgroundColor(ContextCompat.getColor(this@CropImageActivity, R.color.colorAccent))
                }
            } catch (e: FormatException) {
                e.printStackTrace()
                Utils.Log(TAG, "Do not recognize qrcode type FormatException")
                isSaving = false
                isProgressing = false
                btn_done.isEnabled = false
                btn_done.setBackgroundColor(ContextCompat.getColor(this@CropImageActivity, R.color.colorAccent))
            }
        }
    }

    private fun addHint() : MutableMap<DecodeHintType, Any>{
        val tmpHintsMap: MutableMap<DecodeHintType, Any> = EnumMap(
            DecodeHintType::class.java
        )
        tmpHintsMap[DecodeHintType.TRY_HARDER] = java.lang.Boolean.TRUE
        tmpHintsMap[DecodeHintType.POSSIBLE_FORMATS] = EnumSet.allOf(BarcodeFormat::class.java)
        tmpHintsMap[DecodeHintType.PURE_BARCODE] = java.lang.Boolean.TRUE
        return tmpHintsMap
    }

    private fun bitmapRecognize(input : InputStream) : BitmapRegionDecoder?{
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU){
            BitmapRegionDecoder.newInstance(input)
        }else{
            BitmapRegionDecoder.newInstance(input,false)
        }
    }

    companion object {
        private val TAG = CropImageActivity::class.java.simpleName
        private const val SIZE_DEFAULT = 2048
        private const val SIZE_LIMIT = 4096
    }
}